R/toolGetMapping.R

```diff
@@ -24,58 +24,87 @@
-    fname <- paste0(mf, "/", type, "/", name)
-    if (file.exists(as.character(name))) {
-      fname <- name
-    } else if (!file.exists(as.character(fname))) {
-      packages <- getConfig("packages")
-      if (!is.null(activecalc[[1]]) & any(grepl(paste0("^", activecalc[[1]], "$"), getCalculations()[, "type"]))) {
-        fp <- as.character(attr(prepFunctionName(activecalc[[1]], "calc"), "package"))
-        packages <- c(fp, grep(fp, packages, invert = TRUE, value = TRUE))
-      }
-      for (i in packages) {
-        out <- toolGetMapping(name, type = type, where = i, error.missing = FALSE, returnPathOnly = TRUE)
-        if (out != "") {
-          fname <- out
-          break
-        }
+    fname <- .searchNameLocal(name, type = type)
+    if (file.exists(as.character(fname))) {
+      return(fname)
+    }
+    fname <- .searchNameMappingFolder(name, type = type)
+    if (file.exists(as.character(fname))) {
+      return(fname)
+    }
+
+    packages <- getConfig("packages")
+    if (!is.null(activecalc[[1]]) && any(grepl(paste0("^", activecalc[[1]], "$"), getCalculations()[, "type"]))) {
+      fp <- as.character(attr(prepFunctionName(activecalc[[1]], "calc"), "package"))
+      packages <- c(fp, grep(fp, packages, invert = TRUE, value = TRUE))
+    }
+    for (i in packages) {
+      out <- .searchNamePackage(name, type = type, packageName = i)
+      if (out != "") {
+        fname <- out
       }
     }
+    # if not found in packages, fall back to result from mappingfolder, even though it doesn't exist
+    return(fname)
+
   } else if (where == "mappingfolder") {
-    mf <- getConfig("mappingfolder")
-    fname <- paste0(mf, "/", type, "/", name)
+    return(.searchNameMappingFolder(name = name, type = type))
   } else if (where == "local") {
-    if (is.null(type)) {
-      fname <- name
-    } else {
-      fname <- paste0(type, "/", name)
-    }
+    return(.searchNameLocal(name = name, type = type))
   } else {
-    if (is.null(type)) {
-      tmpfname <- name
-    } else {
-      tmpfname <- paste0(type, "/", name)
-    }
-    fname <-  system.file("extdata", tmpfname, package = where)
-    if (fname == "") fname <- system.file("inst/extdata", tmpfname, package = where)
-    if (fname == "") fname <- system.file("extdata", strsplit(tmpfname, split = "/")[[1]][2], package = where)
-    if (fname == "") fname <- system.file("inst/extdata", strsplit(tmpfname, split = "/")[[1]][2], package = where)
-    if (fname == "" & error.missing) {
-      stop('Mapping "', name, '" with type "', type, '" not found in package "', where, '"!')
-    }
+    return(.searchNamePackage(name = name, type = type, packageName = where))
   }
-  if (error.missing & !file.exists(as.character(fname))) stop('Mapping "', name, '" not found!')
-  fname <- gsub("/+", "/", fname)
-  if (returnPathOnly) return(fname)
+}
+
+.searchNameLocal <- function(name, type) {
+  if (file.exists(as.character(name))) {
+    return(name)
+  }
```

Comment:
This check if a bare `name` exists used to be only present in `where = NULL`, not in `where = "local"`, which also meant that the documentation was wrong (which claimed that `where = NULL` starts by doing the same thing as "local"). I think it is somewhat useful to find a mapping if the full path is given and the type is specified, so I added it to `where = "local"` calls as well.